### PR TITLE
proto3 standard compatibility improvements

### PR DESCRIPTION
**Pull request by tim2CF**
- Added `DotProtoMessageOption` constructor to support options inside messages, provided example, included to tests, [example](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-ddaa09b989531549ec6d9c1c28a7697f81aea30dd721cdf0395cfdddd0048d89R14)
- Added `globalIdentifier` parser to support leading dots on field types, provided example, included to tests, [example](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-9a077b7ccec787db0f0efa7e30ebd1701658dbad87a517ff8190dd6ce84f7fa5R14)
- Fixed `SwaggerObject` conditional compilation bug [here](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-11145a54efef3abe0aa32e9547f6bc36daa96f9a659ce14a988c492a8ebe567fR86)
- Fixed shell.nix (cabal might not like wget installed on host machine) [here](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-e53dfbfffe62ae3c0b411b3938ccffa9fb6a2ecc565f55785ef8daa756631a6bR9)
- Added CircleCI support 
- Added proposal related to protoc plugins [here](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-ddaa09b989531549ec6d9c1c28a7697f81aea30dd721cdf0395cfdddd0048d89R5-R8)
- Need help with [this](https://github.com/awakesecurity/proto3-suite/compare/master...coingaming:master#diff-06ec41200d81d80b6460b7a47265101b6c4e7217cd2e8695a8458d6e0a84b2feR129-R130)

### CODE REVIEW

File: tools/canonicalize-proto-file/Main.hs [hunk 1]
```diff
@@ -126,6 +126,8 @@ instance CanonicalRank DotProtoMessagePart
     DotProtoMessageReserved _fs -> Left (Right ())
       -- We use '()' here because 'Canonicalize [DotProtoMessagePart]'
       -- collapses all of the 'DotProtoMessageReserved's into just one.
+    DotProtoMessageOption _ -> Left (Right ())
```
Comment (Gabriel439) [inline tools/canonicalize-proto-file/Main.hs:129]: So I'm new to this section of the code, too, but I think I know what we need to do here.  Basically, this module is formatting a `.proto` file with everything in a specified order.  The way that it works is that it maps each node in the syntax tree to some type that has an `Ord` instance and then uses that `Ord` instance to sort them.

This part of the code would probably be clearer and easier to fix if you first make a small refactor to replace the weird nested `Either` type with a new separate datatype, like this:

```haskell
data PartRank
    = PartRankDefinition (Int, DotProtoIdentifier)
    | PartRankReserved
    | PartRankField (Maybe FieldNumber)

instance CanonicalRank DotProtoMessagePart where
    canonicalRank = \case
      DotProtoMessageField f -> PartRankField (canonicalRank f)
      DotProtoMessageOneOf _ fs -> PartRankField (canonicalRank f)
      DotProtoMessageDefinition d -> PartRankDefinition (canonicalRank d)
      DotProtoMessageReserved _ -> PartRankReserved 
```

… and then once you refactor it like that then it's easier to add in a new constructor for message options, like this:

```haskell
data PartRank
    = PartRankDefinition (Int, DotProtoIdentifier)
    | PartRankReserved
    | PartRankField (Maybe FieldNumber)
    | PartRankOption DotProtoOption

instance CanonicalRank DotProtoMessagePart where
    canonicalRank = \case
      DotProtoMessageField f -> PartRankField (canonicalRank f)
      DotProtoMessageOneOf _ fs -> PartRankField (canonicalRank f)
      DotProtoMessageDefinition d -> PartRankDefinition (canonicalRank d)
      DotProtoMessageReserved _ -> PartRankReserved 
      DotProtoMessageOption o -> PartRankOption o
```

I haven't type-checked that suggestion, so there might be a few mistakes or other things that would need to be fixed, but I think that's the general idea of how to go about this.

**Review by Gabriel439 — COMMENTED** [review]
Looks great!  Just one suggestion

**Review by evanrelf — COMMENTED** [review]
I don't understand the "Added CircleCI support" part of this pull request, since we don't use CircleCI. What's your intention adding `.circleci/config.yml`?

**Comment by tkachuk-labs** [comment]
> I don't understand the "Added CircleCI support" part of this pull request, since we don't use CircleCI. What's your intention adding `.circleci/config.yml`?

We are using CircleCI, in our fork, you can see it there https://github.com/coingaming/proto3-suite/commits/master
In general, CirecleCI is quite popular and working CircleCI config might be useful for people who fork this library, and want CI to be working from the box.

But if you don't want it - I can remove it for sure.

**Comment by evanrelf** [comment]
@tkachuk-labs I think it would be best if you removed it from the pull request to avoid confusion. I don't think it's common for open source projects to have multiple CI configs unless they're actually using those CI services.

**Comment by tkachuk-labs** [comment]
@evanrelf sure, no problem!

**Comment by tim2CF** [comment]
@evanrelf done in https://github.com/awakesecurity/proto3-suite/pull/143/commits/699277b85b5b449120e87c489706aa2391ce6c91

File: tools/canonicalize-proto-file/Main.hs (hunk 1 above)
Comment (tim2CF) [reply tools/canonicalize-proto-file/Main.hs]: @Gabriel439 thanks for advise, it worked perfectly! Fixed in https://github.com/awakesecurity/proto3-suite/pull/143/commits/699277b85b5b449120e87c489706aa2391ce6c91

I also changed order of `PartRank` constructors to provide more logical ordering of protobuf AST.

File: tools/canonicalize-proto-file/Main.hs [hunk 2]
```diff
@@ -116,16 +122,15 @@ instance Canonicalize [DotProtoMessagePart] where
...
   canonicalRank = \case
-    DotProtoMessageField f -> Right (canonicalRank f)
-    DotProtoMessageOneOf _ fs -> Right (canonicalRank fs)
-    DotProtoMessageDefinition d -> Left (Left (canonicalRank d))
-    DotProtoMessageReserved _fs -> Left (Right ())
-      -- We use '()' here because 'Canonicalize [DotProtoMessagePart]'
+    DotProtoMessageOption x -> PartRankOption x
+    DotProtoMessageDefinition d -> PartRankDefinition (canonicalRank d)
+    DotProtoMessageReserved _fs -> PartRankReserved
+      -- We don't use '_fs' here because 'Canonicalize [DotProtoMessagePart]'
```
Comment (Gabriel439) [inline tools/canonicalize-proto-file/Main.hs:130]: This comment needs to be relocated above the `DotProtoMessageOneOf` case after your change

**Review by Gabriel439 — APPROVED** [review]
Looks great!  Just one last thing

File: tools/canonicalize-proto-file/Main.hs (hunk 2 above)
Comment (tim2CF) [reply tools/canonicalize-proto-file/Main.hs:130]: But it was like this before. Comment is related to `DotProtoMessageReserved _fs` AST node. So it depends on style of codebase - should I put comment after related to comment line, or before.

https://github.com/awakesecurity/proto3-suite/blob/c40e73eb9a10e32d033aea3e2c8335af6c4c0926/tools/canonicalize-proto-file/Main.hs#L127-L128

**Comment by tim2CF** [comment]
I did added `swagger2isAbove_2_4_0` optional flag [here](https://github.com/awakesecurity/proto3-suite/pull/143/files#diff-d5f7a586debedc022115cf3432285b544ed70a8d9f2c99634b4cb376ecb939d2R18). It's needed to be able to get suitable version of `proto3-suite.proto3-suite-linux` executable, which can properly generate `hs` files from `proto` files according version of `swagger2` used in application project. By default, before it always was generating `hs` files for swagger2 version > 2.4.0, and this caused compilation errors in generated code. Another possible solution is to move CPP conditional compilation to generated code itself, but it seems harder to do.

```
     * Couldn't match expected type `HsJSONPB.SwaggerType                                 
                                       'swagger2-2.3.1.1:Data.Swagger.Internal.           
 SwaggerKindSchema                                                                        
 '                                                                                        
                   with actual type `Hs.Maybe                                             
                                       (HsJSONPB.SwaggerType                              
                                          'swagger2-2.3.1.1:Data.Swagger.Internal.        
 SwaggerKindSch                                                                           
 ema)'                                                                                    
     * In the `_paramSchemaType' field of a record                                        
       In the `_schemaParamSchema' field of a record                                      
       In the `_namedSchemaSchema' field of a record                                      
     |                                                                                    
 134 |                                                                  Hs.Just HsJSONPB. 
 Swagger                                                                                  
 Object},                                                                                 
     |                                                                                    
 ^^^^^^^^^^^^^^^^^^^^^^^^                                                                 
 ^^^^^^                                                                                   
```

**Comment by Gabriel439** [comment]
@tim2CF: I added you as a collaborator so that you can merge this whenever you think this is ready

**Comment by tim2CF** [comment]
Thanks, Gabriel!

**Comment by Gabriel439** [comment]
You're welcome! 🙂

**Comment by evanrelf** [comment]
@tim2CF I made some non-trivial changes to the Nix code in another PR (https://github.com/awakesecurity/proto3-suite/pull/159) which conflict with the changes you made here. Would you mind if I pushed to your branch to fix up the Nix stuff?

After that, unless anything else is left to be done, I think this'll be ready to merge!

**Comment by tim2CF** [comment]
> @tim2CF I made some non-trivial changes to the Nix code in another PR (#159) which conflict with the changes you made here. Would you mind if I pushed to your branch to fix up the Nix stuff?
> 
> After that, unless anything else is left to be done, I think this'll be ready to merge!

Sure @evanrelf ! I've got busy with the other stuff, so if you will help me to fix conflicts and finally merge, this will be super cool!

**Comment by evanrelf** [comment]
Hmm... @tim2CF, can you see if this option is enabled in the pull request sidebar?

> 

I can't seem to push to `coingaming/proto3-suite`'s `master` branch:

```
ERROR: Permission to coingaming/proto3-suite.git denied to evanrelf.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Or maybe I've misunderstood that GitHub option?

**Comment by tim2CF** [comment]
@evanrelf seems like organization github accounts don't have this feature, but I did invited you as a contributor directly from our fork repo. Please try it

**Comment by evanrelf** [comment]
> I invited you as a contributor directly from our fork repo

Thanks, that worked. This branch is now updated with the latest changes from `master`.

@tim2CF Since Gabriel gave you the commit bit, feel free to merge whenever you're ready!

**Comment by tim2CF** [comment]
Cool, thanks for the help @evanrelf !

**Pull request merged**